sys/include/usb/dfu.h

```diff
@@ -0,0 +1,138 @@
+ *
+ * @author      Dylan Laduranty <dylan.laduranty@mesotic.com>
+ */
+
+#ifndef USB_DFU_H
+#define USB_DFU_H
+
+#include <stdint.h>
+
+#include "usb.h"
+#include "usb/descriptor.h"
+#include "usb/usbus.h"
+#include "usb/usbus/control.h"
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+#define USB_IF_DESCRIPTOR_DFU           0x21   /**< USB DFU type descriptor*/
+#define USB_DFU_VERSION_BCD             0x0110 /**< USB DFU version in BCD */
+
+/**
+ * @name    Default USB detach timeout for DFU descriptor
+ * @{
+ */
+#ifndef USB_DFU_DETACH_TIMEOUT_MS
+#define USB_DFU_DETACH_TIMEOUT_MS 255
+#endif
+/** @} */
+
+/**
+ * @name USB DFU interface attributes
+ * @{
+ */
+#define USB_DFU_CAN_DOWNLOAD            0x01    /**< DFU Download attribute */
+#define USB_DFU_CAN_UPLOAD              0x02    /**< DFU Upload attribute */
+#define USB_DFU_MANIFEST_TOLERANT       0x04    /**< DFU Manifest tolerant attribute */
+#define USB_DFU_WILL_DETACH             0x08    /**< DFU Detach capabability attribute */
+/** @} */
+
+/**
+ * @name USB DFU interface type
+ * @{
+ */
+#define USB_DFU_INTERFACE               0xFE /** Application Specific Interface */
+/** @} */
+
+/**
+ * @name USB DFU subclass types
+ * @anchor usb_dfu_subtype
+ * @{
+ */
+#define USB_DFU_SUBCLASS_DFU            0x01 /**< DFU subclass */
+/** @} */
+
+/**
+ * @name USB DFU protocol types
+ * @{
+ */
+#define USB_DFU_PROTOCOL_RUNTIME_MODE   0x01 /**< Runtime mode */
+#define USB_DFU_PROTOCOL_DFU_MODE       0x02 /**< DFU mode */
+/** @} */
+
+/**
+ * @name USB DFU setup request
+ * @{
+ */
+#define DFU_DETACH      0x00        /**< DFU Detach request */
+#define DFU_DOWNLOAD    0x01        /**< DFU Download request */
+#define DFU_UPLOAD      0x02        /**< DFU Upload request */
+#define DFU_GET_STATUS  0x03        /**< DFU Get Status request */
+#define DFU_CLR_STATUS  0x04        /**< DFU Clear Status request */
+#define DFU_GET_STATE   0x05        /**< DFU Get State request */
+#define DFU_ABORT       0x06        /**< DFU Abort request */
+/** @} */
+
+/**
+ * @brief USBUS DFU internal state
+ */
+typedef enum {
+    USBUS_DFU_STATE_APP_IDLE,               /**< DFU application idle */
+    USBUS_DFU_STATE_APP_DETACH,             /**< DFU application detach (reboot to DFU mode) */
+    USBUS_DFU_STATE_DFU_IDLE,               /**< DFU runtime mode idle */
+    USBUS_DFU_STATE_DFU_DL_SYNC,            /**< DFU download synchronization */
+    USBUS_DFU_STATE_DFU_DL_BUSY,            /**< DFU download busy */
+    USBUS_DFU_STATE_DFU_DL_IDLE,            /**< DFU download idle */
+    USBUS_DFU_STATE_DFU_MANIFEST_SYNC,      /**< DFU manifest synchronization */
+    USBUS_DFU_STATE_DFU_MANIFEST,           /**< DFU manifest mode */
+    USBUS_DFU_STATE_DFU_MANIFEST_WAIT_RST,  /**< DFU manifest wait for CPU reset */
+    USBUS_DFU_STATE_DFU_UP_IDLE,            /**< DFU upload idle */
+    USBUS_DFU_STATE_DFU_ERROR               /**< DFU internal error */
+} usbus_dfu_state_t;
```

Comment:
This should be moved to `sys/include/usb/usbus/dfu.h` if this is USBUS-specific.

Reply:
It is not USBUS-specific but part of the DFU spec. These values are used for an internal state machine and the host can access it by issuing a GET_STATUS request.
I can rename them to USB_DFU_STATE_* instead of USBUS_DFU_STATE_* if you prefer.

Reply:
>  I can rename them to USB_DFU_STATE_* instead of USBUS_DFU_STATE_* if you prefer.

Sounds good!

Reply:
Done.